docs/proposals/cluster-mesh.md

```diff
@@ -0,0 +1,101 @@
+
+## Non-Goals
+We do not focus on the networking datapath, or stitching the cross-cluster connectivity ourselves, instead we solely rely on the existing Cilium Cluster Mesh feature with all its benefits and limitations.
+
+## Motivation and Background
+
+The main motivation for interconnecting multiple clusters are the following use-cases:
+
+- High availability of clusters (failover of a service to a different cluster, possibly in a different datacenter),
+- Easy workload migration between clusters,
+- Hybrid cloud scenarios, like providing special services by on-premises clusters to cloud clusters, or vice-versa.
+
+This should be in the ideal case possible without the need for too complex setup on the cluster level and service level (e.g. with automatic service discovery), and in a secure manner (e.g. network policies should work across multiple clusters, traffic between the clusters can be encrypted).
+
+All of this can be achieved by relying on the feature of the Cilium CNI called Cluster Mesh.
+
+## Implementation
+
+We will add a new CRD on the Seed level (`ClusterMesh`), that would define a list of clusters that are part of the mesh.
+
+In the initial implementation it will be possible to mesh only clusters that belong to the same KKP seed cluster, to avoid flow of data from the user clusters / KKP seed clusters towards the KKP main cluster. This can be enhanced without major changes in the future, e.g. if Cilium Cluster Mesh deployment will be simplified.
+
+The `ClusterMesh` CRD will be defined as follows:
+
+```go
+// ClusterMesh is the object representing a mesh of multiple KKP user clusters.
+type ClusterMesh struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  Spec ClusterMeshSpec `json:"spec,omitempty"`
+}
+
+// ClusterMeshSpec specifies the mesh of KKP user clusters.
+type ClusterMeshSpec struct {
+  // Clusters is a list of clusters connected in the ClusterMesh.
+  Clusters []ClusterMeshCluster `json:"clusters,omitempty"`
+}
+
+// ClusterMeshCluster contains mesh configuration for a cluster.
+type ClusterMeshCluster struct {
+  // ID an internal identifier of the cluster in the ClusterMesh.
+  // Must be unique across clusters in a ClusterMesh.
+  ID int `json:"ID,omitempty"`
+
+  // APIServerServiceType defines the k8s service type used to expose the clustermesh apiserver the other clusters.
+  APIServerServiceType v1.ServiceType `json:"apiServerServiceType,omitempty"`
+
+  // APIServerIPs contains a list of IPs on which the clustermesh apiserver of this cluster is exposed to the other clusters.
+  // It can be a single IP if APIServerServiceType == LoadBalancer, or multiple IPs if APIServerServiceType == NodePort.
+  APIServerIPs []string `json:"apiServerIPs,omitempty"`
+
+  // APIServerPort defines the port on which the clustermesh apiserver of this cluster is exposed to the other clusters.
+  APIServerPort int `json:"apiServerPort,omitempty"`
+}
```

Comment:
how would I reference a KKP user cluster from this struct? As far as I understood the other documentation, you create one `ClusterMesh` resource for X clusters you want to interconnect, but I don't see a way to actually list the X clusters I want.

Reply:
Damn, OFC, I completely forgot to include the KKP cluster name in the struct